### ISSUE TITLE
Rack::Test: Undefined method `app` when using Rails app

### ISSUE THREAD

**Issue by danascheider** [danascheider/rambo#25]
When running Rambo against a Rails app, all the generated tests fail with the following message:
```
  8) Repaymnt Estimatr /max_project_amount POST returns status 200
     Failure/Error: post route

     NameError:
       undefined local variable or method `app' for #<RSpec::ExampleGroups::RepaymntEstimatr::MaxProjectAmount::POST:0x007f9fa7c4de68>
     # /Users/dscheider/.rvm/gems/ruby-2.1.7@repaymnt_estimatr/gems/rack-test-0.6.3/lib/rack/test/methods.rb:31:in `build_rack_mock_session'
     # /Users/dscheider/.rvm/gems/ruby-2.1.7@repaymnt_estimatr/gems/rack-test-0.6.3/lib/rack/test/methods.rb:27:in `rack_mock_session'
     # /Users/dscheider/.rvm/gems/ruby-2.1.7@repaymnt_estimatr/gems/rack-test-0.6.3/lib/rack/test/methods.rb:42:in `build_rack_test_session'
     # /Users/dscheider/.rvm/gems/ruby-2.1.7@repaymnt_estimatr/gems/rack-test-0.6.3/lib/rack/test/methods.rb:38:in `rack_test_session'
     # /Users/dscheider/.rvm/gems/ruby-2.1.7@repaymnt_estimatr/gems/rack-test-0.6.3/lib/rack/test/methods.rb:46:in `current_session'
     # ./spec/contract/repaymnt_estimatr_spec.rb:1007:in `block (4 levels) in <top (required)>'
```
The problem here is that Rack::Test requires an `#app` method to be defined. In this case, the code that solved this problem when added manually to the contract test file was this:
```ruby
describe "Repaymnt Estimatr" do 
  include Rack::Test::Methods

  def app
    RepaymntEstimatr::Application
  end
  ...
end
```
There are a couple of issues here. 

Firstly, Rambo has no way of knowing the class of the app. Furthermore, finding out the class of the app differs substantially by framework. Once the class of the app is known, there is the question of whether it needs to be instantiated in the `app` method. For instance, whereas a Rails app will be included as above, a Sinatra app might look more like this:
```ruby
def app
  SinatraApp.new
end
```
### Tasks
There are a few action items/problems to solve associated with this issue.

#### Scope of Issue
So far, Rambo has only been run against a Rails app, Repaymnt Estimatr, which is being used to guide development (sorry OSS contributors - it's private 😞). Consequently, it is unknown whether this problem even occurs with other frameworks. We need to find out whether Rack::Test always requires an `app` method, and if not, under what circumstances this is required. If this is an issue with frameworks other than Rails, then the issue should be addressed for all supported frameworks, which brings us to...

#### Supported Frameworks
As a Ruby tool designed to test web apps, Rails support is clearly _de rigueur_. I would also expect Sinatra and Rack itself to be supported. As for other frameworks, we'll need to determine whether there is a way to easily support all Rack-based frameworks and, if not, which other frameworks, if any, should be supported. Padrino would be the only additional candidate I think is widely enough used to warrant particular consideration here, and non-Rack-based frameworks, which I'm sure exist, would not be a priority at all.

#### Implementation
The naïve implementation here would be to find the appropriate `.rb` file (e.g., `/config/application.rb`) and identify the name of the class from there. However, I'm wondering whether there is a more elegant - and more importantly, foolproof - way of doing this.

**Comment by dolan** [comment]
You could possibly use a railtie to load whatever you need for rails. And just assume that it's not rails if it isn't loaded.

**Comment by danascheider** [comment]
Good idea, thanks @dolan!

**Comment by danascheider** [comment]
Resolved in #31
